src/Raven.Server/Documents/Revisions/RevisionsStorage.cs

```diff
@@ -1898,7 +1898,40 @@ public void ForceDeleteAllRevisionsFor(DocumentsOperationContext context, string
                 (table, result) => GetAllRevisions(context, table, prefixSlice, maxDeletesUponUpdate, shouldSkip, result), tombstoneFlags);
         }
 
-        private (bool MoreWork, long Deleted) ForceDeleteAllRevisionsFor(DocumentsOperationContext context, Slice lowerId, Slice prefixSlice, CollectionName collectionName, 
+        private (bool MoreWork, long Deleted) ForceDeleteAllRevisionsFor(DocumentsOperationContext context, Slice lowerId, Slice prefixSlice, CollectionName collectionName, long? maxDeletesUponUpdate,
+            long etagBarrier, DocumentFlags tombstoneFlags = DocumentFlags.None)
+        {
+            return ForceDeleteAllRevisionsFor(context, lowerId, prefixSlice, collectionName, GetRevisions, tombstoneFlags);
+
+            IEnumerable<Document> GetRevisions(Table table, DeleteOldRevisionsResult result)
+            {
+                var revisions = GetAllRevisions(context, table, prefixSlice, maxDeletesUponUpdate, shouldSkip: null, result);
+                foreach (var r in revisions)
+                {
+                    if (context.CanContinueTransaction == false)
+                    {
+                        result.HasMore = true;
+                        yield break;
+                    }
+
+                    var etag = r.Etag;
+                    if (etag > etagBarrier)
+                    {
+                        context.Transaction.ForgetAbout(r);
+                        r.Dispose();
+                        yield break;
+                    }
+            
+                    yield return r;
+
+                    if (etag == etagBarrier) // for not stopping on maxDeletesUponUpdate instead of etagBarrier
+                        yield break;
```

Comment:
not sure I follow this...

Reply:
We talked about that and the edge cases it prevents